File: sdk/go/worker.go
package tako

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"math"
	"math/rand"
	"os"
	"sync"
	"sync/atomic"
	"time"
)

// Handler is the function signature a workflow handler must satisfy.
type Handler func(ctx *WorkflowContext, payload json.RawMessage) error

// WorkflowContext is passed to handlers during execution.
type WorkflowContext struct {
	// RunID is the unique id for the current workflow run.
	RunID string
	// WorkflowName is the registered workflow name.
	WorkflowName string
	// Attempts is the number of attempts already made for this run.
	Attempts uint32
	// Step exposes checkpointed step helpers for this run.
	Step *StepAPI
	ctx  context.Context
}

// Done returns the context's cancellation channel.
func (c *WorkflowContext) Done() <-chan struct{} { return c.ctx.Done() }

// Bail returns a sentinel error. Returning it ends the run cleanly as
// `cancelled` (no retries). To exit successfully early, just return nil
// from the handler.
func (c *WorkflowContext) Bail(reason string) error {
	return &bailSignal{reason: reason}
}

// Fail returns a sentinel error. Returning it marks the run `dead`
// immediately, skipping any remaining retry budget.
func (c *WorkflowContext) Fail(err error) error {
	if err == nil {
		err = errors.New("tako: ctx.Fail with nil error")
	}
	return &failSignal{err: err}
}

type bailSignal struct{ reason string }

func (b *bailSignal) Error() string { return "tako: bail: " + b.reason }

type failSignal struct{ err error }

func (f *failSignal) Error() string { return "tako: fail: " + f.err.Error() }
func (f *failSignal) Unwrap() error { return f.err }

type deferSignal struct{ wakeAt *time.Time }

func (d *deferSignal) Error() string { return "tako: defer signal" }

type waitSignal struct {
	stepName  string
	eventName string
	timeoutAt *time.Time
}

func (w *waitSignal) Error() string { return "tako: wait signal" }

// StepAPI is the checkpointed step runner.
type StepAPI struct {
	client   *Client
	ctx      context.Context
	runID    string
	workerID string
	state    map[string]any
	mu       sync.Mutex
}

// StepRunOpts configures a single step.Run invocation.
type StepRunOpts struct {
	// Retries is the in-step retry budget (default 0 = no internal retries).
	Retries uint32
	// Backoff configures in-step retry delays.
	Backoff struct {
		// Base is the initial retry delay. Zero means 1s.
		Base time.Duration
		// Max is the retry delay cap. Zero means 30s.
		Max time.Duration
	}
	// NoRetry fails the run immediately on any fn error, skipping in-step and run-level retries.
	NoRetry bool
}

const inlineSleepThreshold = 30 * time.Second

// Run executes fn the first time and persists its return value. On retry
// the stored value is returned without calling fn.
//
// At-least-once: fn may run more than once on crashes — make it
// idempotent (Stripe idempotency keys, upsert not insert, etc.).
func (s *StepAPI) Run(name string, out any, fn func() (any, error), opts ...StepRunOpts) error {
	s.mu.Lock()
	if cached, ok := s.state[name]; ok {
		s.mu.Unlock()
		return assignCached(cached, out)
	}
	s.mu.Unlock()

	var o StepRunOpts
	if len(opts) > 0 {
		o = opts[0]
	}
	base := o.Backoff.Base
	if base == 0 {
		base = time.Second
	}
	max := o.Backoff.Max
	if max == 0 {
		max = 30 * time.Second
	}
	attempts := o.Retries + 1

	var lastErr error
	for attempt := uint32(1); attempt <= attempts; attempt++ {
		value, err := fn()
		if err == nil {
			s.mu.Lock()
			s.state[name] = value
			s.mu.Unlock()
			if perr := s.client.SaveStep(s.ctx, s.runID, s.workerID, name, value); perr != nil {
				return fmt.Errorf("persist step state: %w", perr)
			}
			return assignCached(value, out)
		}
		// Control signals propagate immediately — never retry them.
		if isControlSignal(err) {
			return err
		}
		lastErr = err
		if o.NoRetry {
			return &failSignal{err: err}
		}
		if attempt < attempts {
			time.Sleep(expBackoff(attempt, base, max))
		}
	}
	return lastErr
}

func isControlSignal(err error) bool {
	var bs *bailSignal
	var fs *failSignal
	var ds *deferSignal
	var ws *waitSignal
	return errors.As(err, &bs) || errors.As(err, &fs) ||
		errors.As(err, &ds) || errors.As(err, &ws)
}

// Sleep durably waits for d. Short sleeps run inline; longer sleeps defer
// the run via the server so the worker process can release.
func (s *StepAPI) Sleep(name string, d time.Duration) error {
	key := "__sleep:" + name
	s.mu.Lock()
	if stored, ok := s.state[key]; ok {
		s.mu.Unlock()
		if m, ok := stored.(map[string]any); ok {
			if wakeMs, ok := m["wakeAt"].(float64); ok {
				wakeAt := time.UnixMilli(int64(wakeMs))
				if time.Now().Before(wakeAt) {
					return &deferSignal{wakeAt: &wakeAt}
				}
				s.mu.Lock()
				if _, hasName := s.state[name]; !hasName {
					s.state[name] = true
					s.mu.Unlock()
					return s.client.SaveStep(s.ctx, s.runID, s.workerID, name, true)
				}
				s.mu.Unlock()
				return nil
			}
		}
		return nil
	}
	s.mu.Unlock()

	wakeAt := time.Now().Add(d)
	wakeMs := wakeAt.UnixMilli()
	marker := map[string]any{"wakeAt": wakeMs}
	s.mu.Lock()
	s.state[key] = marker
	s.mu.Unlock()
	if err := s.client.SaveStep(s.ctx, s.runID, s.workerID, key, marker); err != nil {
		return err
	}
	if d < inlineSleepThreshold {
		time.Sleep(d)
		s.mu.Lock()
		s.state[name] = true
		s.mu.Unlock()
		return s.client.SaveStep(s.ctx, s.runID, s.workerID, name, true)
	}
	return &deferSignal{wakeAt: &wakeAt}
}

// WaitFor parks the run waiting for a Signal with the given name. timeout
// of 0 means "no timeout" (parked indefinitely until matching signal). On
// resume, out (a *T) is populated with the signal's payload.
func (s *StepAPI) WaitFor(name string, out any, timeout time.Duration) error {
	s.mu.Lock()
	cached, ok := s.state[name]
	s.mu.Unlock()
	if ok {
		return assignCached(cached, out)
	}
	var timeoutAt *time.Time
	if timeout > 0 {
		t := time.Now().Add(timeout)
		timeoutAt = &t
	}
	return &waitSignal{stepName: name, eventName: name, timeoutAt: timeoutAt}
}

func assignCached(cached, out any) error {
	if out == nil {
		return nil
	}
	bs, err := json.Marshal(cached)
	if err != nil {
		return err
	}
	return json.Unmarshal(bs, out)
}

type workflowRegistration struct {
	handler Handler
	config  workflowConfig
}

type workflowConfig struct {
	maxAttempts uint32
	timeoutMs   uint32
	schedule    string
	backoffBase time.Duration
	backoffMax  time.Duration
}

// WorkflowOption configures a registered workflow.
type WorkflowOption func(*workflowConfig)

// WithMaxAttempts sets the total run-level attempt budget.
func WithMaxAttempts(n uint32) WorkflowOption {
	return func(c *workflowConfig) { c.maxAttempts = n }
}

// WithSchedule registers a cron schedule for this workflow.
func WithSchedule(expr string) WorkflowOption {
	return func(c *workflowConfig) { c.schedule = expr }
}

// WithBackoff sets run-level backoff between failed attempts.
func WithBackoff(base, max time.Duration) WorkflowOption {
	return func(c *workflowConfig) {
		c.backoffBase = base
		c.backoffMax = max
	}
}

var (
	registryMu      sync.Mutex
	registry        = map[string]workflowRegistration{}
	registryStarted bool
)

// RegisterWorkflow registers a handler under `name`. Must be called before
// RunWorker.
func RegisterWorkflow(name string, h Handler, opts ...WorkflowOption) {
	registryMu.Lock()
	defer registryMu.Unlock()
	if registryStarted {
		panic(fmt.Sprintf("tako: RegisterWorkflow(%q) called after RunWorker started", name))
	}
	if _, dup := registry[name]; dup {
		panic(fmt.Sprintf("tako: workflow %q already registered", name))
	}
	var cfg workflowConfig
	for _, opt := range opts {
		opt(&cfg)
	}
	registry[name] = workflowRegistration{handler: h, config: cfg}
}

// RunWorker connects to the enqueue socket, registers schedules, and runs
// the claim loop until ctx is cancelled or the idle timeout fires.
func RunWorker(ctx context.Context) error {
	registryMu.Lock()
	if len(registry) == 0 {
		registryMu.Unlock()
		return errors.New("tako: no workflows registered")
	}
	registryStarted = true
	handlers := make(map[string]workflowRegistration, len(registry))
	for k, v := range registry {
		handlers[k] = v
	}
	registryMu.Unlock()

	client, err := ClientFromEnv()
	if err != nil {
		return err
	}

	var schedules []ScheduleSpec
	for name, reg := range handlers {
		if reg.config.schedule != "" {
			schedules = append(schedules, ScheduleSpec{Name: name, Cron: reg.config.schedule})
		}
	}
	if len(schedules) > 0 {
		if err := client.RegisterSchedules(ctx, schedules); err != nil {
			return fmt.Errorf("register schedules: %w", err)
		}
	}

	w := newWorker(ctx, client, handlers)
	return w.run()
}

type worker struct {
	ctx         context.Context
	client      *Client
	handlers    map[string]workflowRegistration
	workerID    string
	leaseMs     uint64
	heartbeatMs uint64
	pollMs      int
	baseBackoff time.Duration
	maxBackoff  time.Duration
	idleTimeout time.Duration
	lastClaimAt atomic.Int64
}

func newWorker(ctx context.Context, client *Client, handlers map[string]workflowRegistration) *worker {
	w := &worker{
		ctx:         ctx,
		client:      client,
		handlers:    handlers,
		workerID:    fmt.Sprintf("worker-%d", os.Getpid()),
		leaseMs:     60_000,
		heartbeatMs: 20_000,
		pollMs:      1_000,
		baseBackoff: time.Second,
		maxBackoff:  time.Hour,
		idleTimeout: parseIdleTimeout(),
	}
	w.lastClaimAt.Store(time.Now().UnixMilli())
	return w
}

func (w *worker) run() error {
	names := make([]string, 0, len(w.handlers))
	for name := range w.handlers {
		names = append(names, name)
	}

	for {
		select {
		case <-w.ctx.Done():
			return nil
		default:
		}

		task, err := w.client.Claim(w.ctx, w.workerID, names, w.leaseMs)
		if err != nil {
			fmt.Fprintln(os.Stderr, "tako-worker: claim error:", err)
			time.Sleep(time.Duration(w.pollMs) * time.Millisecond)
			continue
		}
		if task == nil {
			if w.idleTimeout > 0 &&
				time.Since(time.UnixMilli(w.lastClaimAt.Load())) >= w.idleTimeout {
				return nil
			}
			select {
			case <-w.ctx.Done():
				return nil
			case <-time.After(time.Duration(w.pollMs) * time.Millisecond):
			}
			continue
		}

		w.lastClaimAt.Store(time.Now().UnixMilli())
		w.execute(task)
	}
}

func (w *worker) execute(task *Run) {
	reg, ok := w.handlers[task.Name]
	if !ok {
		_ = w.client.Fail(w.ctx, task.ID, w.workerID, fmt.Sprintf("no handler registered for %q", task.Name), nil, true)
		return
	}

	stepState := task.StepState
	if stepState == nil {
		stepState = map[string]any{}
	}
	ctx := &WorkflowContext{
		RunID:        task.ID,
		WorkflowName: task.Name,
		Attempts:     task.Attempts,
		Step: &StepAPI{
			client:   w.client,
			ctx:      w.ctx,
			runID:    task.ID,
			workerID: w.workerID,
			state:    stepState,
		},
		ctx: w.ctx,
	}

	hbStop := make(chan struct{})
	go w.heartbeatLoop(task.ID, hbStop)
	defer close(hbStop)

	err := reg.handler(ctx, task.Payload)
	if err != nil {
		var bs *bailSignal
		var fs *failSignal
		var ds *deferSignal
		var ws *waitSignal
		switch {
		case errors.As(err, &bs):
			reason := bs.reason
			var rp *string
			if reason != "" {
				rp = &reason
			}
			_ = w.client.Cancel(w.ctx, task.ID, w.workerID, rp)
		case errors.As(err, &fs):
			_ = w.client.Fail(w.ctx, task.ID, w.workerID, fs.err.Error(), nil, true)
		case errors.As(err, &ds):
			_ = w.client.Defer(w.ctx, task.ID, w.workerID, ds.wakeAt)
		case errors.As(err, &ws):
			_ = w.client.WaitForEvent(w.ctx, task.ID, w.workerID, ws.stepName, ws.eventName, ws.timeoutAt)
		default:
			maxAttempts := reg.config.maxAttempts
			if maxAttempts == 0 {
				maxAttempts = task.MaxAttempts
			}
			finalize := task.Attempts >= maxAttempts
			var next *time.Time
			if !finalize {
				base := reg.config.backoffBase
				if base == 0 {
					base = w.baseBackoff
				}
				max := reg.config.backoffMax
				if max == 0 {
					max = w.maxBackoff
				}
				t := time.Now().Add(expBackoff(task.Attempts, base, max))
				next = &t
			}
			_ = w.client.Fail(w.ctx, task.ID, w.workerID, err.Error(), next, finalize)
		}
		return
	}
	_ = w.client.Complete(w.ctx, task.ID, w.workerID)
}

func (w *worker) heartbeatLoop(id string, stop <-chan struct{}) {
	ticker := time.NewTicker(time.Duration(w.heartbeatMs) * time.Millisecond)
	defer ticker.Stop()
	for {
		select {
		case <-stop:
			return
		case <-ticker.C:
			_ = w.client.Heartbeat(w.ctx, id, w.workerID, w.leaseMs)
		}
	}
}

func expBackoff(attempts uint32, base, max time.Duration) time.Duration {
	if attempts < 1 {
		attempts = 1
	}
	exp := base * time.Duration(math.Pow(2, float64(attempts-1)))
	if exp > max {
		exp = max
	}
	jitter := time.Duration(float64(exp) * 0.2 * (rand.Float64()*2 - 1))
	result := exp + jitter
	if result < 0 {
		result = 0
	}
	return result
}

func parseIdleTimeout() time.Duration {
	raw := os.Getenv("TAKO_WORKER_IDLE_TIMEOUT_MS")
	if raw == "" {
		return 0
	}
	var ms int64
	_, err := fmt.Sscanf(raw, "%d", &ms)
	if err != nil || ms < 0 {
		return 0
	}
	return time.Duration(ms) * time.Millisecond
}

File: sdk/javascript/src/workflows/worker.ts
/**
 * Worker loop — claims one run at a time, executes its handler with a
 * checkpointed step API, heartbeats the lease, and finalizes the run via
 * `complete` / `cancel` / `fail` / `defer` / `wait_for_event` based on the
 * outcome.
 *
 * Sentinel exceptions (BailSignal/FailSignal/DeferSignal/WaitSignal) drive
 * the non-retry termination paths cleanly.
 *
 * Start one Worker per app instance. `drain()` stops claiming and awaits
 * any in-flight run for the platform drain hook.
 */

import { createLogger, type Logger } from "../logger";
import { expBackoffMs } from "./backoff";
import type { WorkflowsClient } from "./rpc-client";
import {
  BailSignal,
  createStepAPI,
  DeferSignal,
  FailSignal,
  WaitSignal,
  type StepRunOptions,
  type StepWaitOptions,
  type WorkflowStepContext,
} from "./step";
import type { Run, StepState } from "./types";

/**
 * Function that executes one workflow run.
 *
 * The payload type `P` is inferred from `defineWorkflow<P>(...)` and flows into
 * `.enqueue(payload)`.
 */
export type WorkflowHandler<P = unknown> = (
  payload: P,
  ctx: WorkflowContext,
) => Promise<void> | void;

/**
 * Runtime context passed to each workflow handler.
 */
export interface WorkflowContext {
  /** Unique id for the current workflow run. */
  readonly runId: string;
  /** Name used when the workflow was registered with `defineWorkflow`. */
  readonly workflowName: string;
  /** Current run attempt, starting at 1. */
  readonly attempt: number;
  /** Logger scoped to this workflow run. */
  readonly logger: Logger;
  /**
   * Execute and memoize a durable step.
   *
   * On retry, completed steps return their stored result without calling `fn`.
   */
  run<T>(
    name: string,
    fn: (step: WorkflowStepContext) => Promise<T> | T,
    opts?: StepRunOptions,
  ): Promise<T>;
  /**
   * Durably sleep for `durationMs`.
   */
  sleep(name: string, durationMs: number): Promise<void>;
  /**
   * Park the run until `signal(name, payload)` wakes it or the optional timeout
   * elapses.
   */
  waitFor<T = unknown>(name: string, opts?: StepWaitOptions): Promise<T | null>;
  /**
   * End the run cleanly as `cancelled` with no retries.
   *
   * Use this for work that is no longer needed. To exit successfully early,
   * just `return` from the handler.
   */
  bail(reason?: string): never;
  /**
   * End the run as `dead` immediately with no retries.
   *
   * Use this for permanent errors that will not improve with retry.
   */
  fail(error: Error | string): never;
}

interface WorkflowRetryConfig {
  /** Run-level retry budget (default 3). */
  maxAttempts?: number;
  /** Run-level backoff between failed attempts. */
  backoff?: {
    /**
     * Initial backoff delay in ms.
     * @defaultValue 1_000
     */
    base?: number;
    /**
     * Maximum backoff delay in ms.
     * @defaultValue 3_600_000
     */
    max?: number;
  };
}

/** Options for constructing a workflow worker. */
export interface WorkerOptions {
  /** RPC client used for queue operations. */
  client: WorkflowsClient;
  /** Registered workflow handlers by workflow name. */
  registry: Map<string, RegisteredWorkflow>;
  /** Stable worker id used for leases. */
  workerId: string;
  /** @defaultValue 60_000 */
  leaseMs?: number;
  /** @defaultValue leaseMs / 3 */
  heartbeatIntervalMs?: number;
  /** @defaultValue 1_000 */
  pollIntervalMs?: number;
  /** @defaultValue 1_000 */
  baseBackoffMs?: number;
  /** @defaultValue 3_600_000 */
  maxBackoffMs?: number;
  /**
   * Scale-to-zero: exit poll loop after this many ms with no claim.
   * @defaultValue 0
   */
  idleTimeoutMs?: number;
  /**
   * Max concurrent in-flight runs.
   * @defaultValue 500
   */
  concurrency?: number;
  /** Base logger. Per-run children are tagged `worker:<workflowName>`. */
  logger?: Logger;
}

/** Workflow handler registered with the worker runtime. */
export interface RegisteredWorkflow {
  /** Function that executes one workflow run. */
  handler: WorkflowHandler;
  /** Optional run-level retry configuration. */
  retry?: WorkflowRetryConfig;
}

const DEFAULTS = {
  leaseMs: 60_000,
  pollIntervalMs: 1_000,
  baseBackoffMs: 1_000,
  maxBackoffMs: 3_600_000,
  idleTimeoutMs: 0,
  concurrency: 500,
} as const;

/**
 * Workflow worker loop.
 *
 * Claims runs from the workflow RPC client, executes handlers with durable
 * step APIs, heartbeats leases, and finalizes run state.
 */
export class Worker {
  private readonly client: WorkflowsClient;
  private readonly registry: Map<string, RegisteredWorkflow>;
  private readonly workerId: string;
  private readonly leaseMs: number;
  private readonly heartbeatIntervalMs: number;
  private readonly pollIntervalMs: number;
  private readonly baseBackoffMs: number;
  private readonly maxBackoffMs: number;
  private readonly idleTimeoutMs: number;
  private readonly concurrency: number;
  private readonly log: Logger;

  private draining = false;
  private idledOut = false;
  private lastClaimAt = 0;
  private readonly inFlight = new Set<Promise<void>>();
  private loopPromise: Promise<void> | null = null;

  constructor(opts: WorkerOptions) {
    this.client = opts.client;
    this.registry = opts.registry;
    this.workerId = opts.workerId;
    this.leaseMs = opts.leaseMs ?? DEFAULTS.leaseMs;
    this.heartbeatIntervalMs = opts.heartbeatIntervalMs ?? Math.floor(this.leaseMs / 3);
    this.pollIntervalMs = opts.pollIntervalMs ?? DEFAULTS.pollIntervalMs;
    this.baseBackoffMs = opts.baseBackoffMs ?? DEFAULTS.baseBackoffMs;
    this.maxBackoffMs = opts.maxBackoffMs ?? DEFAULTS.maxBackoffMs;
    this.idleTimeoutMs = opts.idleTimeoutMs ?? DEFAULTS.idleTimeoutMs;
    this.concurrency = Math.max(1, opts.concurrency ?? DEFAULTS.concurrency);
    this.log = opts.logger ?? createLogger("worker");
    this.lastClaimAt = Date.now();
  }

  /** True when this worker exited because `idleTimeoutMs` elapsed. */
  get idled(): boolean {
    return this.idledOut;
  }

  /** Claim and process one run if available. */
  async processOnce(): Promise<boolean> {
    if (this.draining) return false;
    const names = Array.from(this.registry.keys());
    const run = await this.client.claim(this.workerId, names, this.leaseMs);
    if (!run) return false;
    this.lastClaimAt = Date.now();

    const work = this.execute(run);
    this.inFlight.add(work);
    try {
      await work;
    } finally {
      this.inFlight.delete(work);
    }
    return true;
  }

  /**
   * Claim one run and kick off its execution without awaiting. Returns
   * `true` if a run was claimed and dispatched, `false` if the queue was
   * empty. Used by the internal run loop to keep up to `concurrency`
   * runs in flight simultaneously; tests use the awaiting `processOnce`.
   */
  private async dispatchOne(): Promise<boolean> {
    if (this.draining) return false;
    const names = Array.from(this.registry.keys());
    const run = await this.client.claim(this.workerId, names, this.leaseMs);
    if (!run) return false;
    this.lastClaimAt = Date.now();

    const work: Promise<void> = this.execute(run)
      .catch((err) => {
        // Nothing awaits this promise; without a catch, a failed finalize
        // RPC (complete/fail/cancel) becomes an unhandled rejection.
        this.log.error("run finalization failed", {
          runId: run.id,
          workflow: run.name,
          error: err instanceof Error ? err.message : String(err),
        });
      })
      .finally(() => {
        this.inFlight.delete(work);
      });
    this.inFlight.add(work);
    return true;
  }

  /** Start the polling loop. */
  start(): void {
    if (this.loopPromise) return;
    this.loopPromise = this.runLoop();
  }

  /** Stop claiming new runs and wait for active runs to finish. */
  async drain(): Promise<void> {
    this.draining = true;
    if (this.loopPromise) {
      await this.loopPromise;
      this.loopPromise = null;
    }
    await Promise.allSettled(Array.from(this.inFlight));
  }

  /** Number of currently running workflow handlers. */
  get runningCount(): number {
    return this.inFlight.size;
  }

  private async runLoop(): Promise<void> {
    while (!this.draining) {
      if (this.inFlight.size >= this.concurrency) {
        await Promise.race(Array.from(this.inFlight)).catch(() => {});
        continue;
      }
      const did = await this.dispatchOne().catch(() => false);
      if (!did && !this.draining) {
        if (
          this.idleTimeoutMs > 0 &&
          this.inFlight.size === 0 &&
          Date.now() - this.lastClaimAt >= this.idleTimeoutMs
        ) {
          this.idledOut = true;
          this.draining = true;
          break;
        }
        await new Promise((r) => setTimeout(r, this.pollIntervalMs));
      }
    }
  }

  private async execute(run: Run): Promise<void> {
    const runLog = this.log.child(`worker:${run.name}`, { runId: run.id });
    const contextLog = this.log.child(run.name, { runId: run.id, workflow: run.name });
    const reg = this.registry.get(run.name);
    if (!reg) {
      runLog.error("Workflow failed", { error: `no handler registered for '${run.name}'` });
      await this.client.fail(
        run.id,
        this.workerId,
        `no handler registered for '${run.name}'`,
        null,
        true,
      );
      return;
    }

    const stepState: StepState = { ...run.stepState };
    const createStepContext = (stepName: string): WorkflowStepContext => ({
      runId: run.id,
      workflowName: run.name,
      stepName,
      attempt: run.attempts,
      logger: contextLog.child(`${run.name}:${stepName}`, { step: stepName }),
    });
    const context: WorkflowContext = {
      runId: run.id,
      workflowName: run.name,
      attempt: run.attempts,
      logger: contextLog,
      ...createStepAPI(this.client, run.id, this.workerId, stepState, runLog, createStepContext),
      bail: (reason?: string): never => {
        throw new BailSignal(reason);
      },
      fail: (error: Error | string): never => {
        const e = error instanceof Error ? error : new Error(error);
        throw new FailSignal(e);
      },
    };

    let heartbeatTimer: ReturnType<typeof setInterval> | null = null;
    if (this.heartbeatIntervalMs > 0) {
      heartbeatTimer = setInterval(() => {
        this.client.heartbeat(run.id, this.workerId, this.leaseMs).catch(() => {});
      }, this.heartbeatIntervalMs);
    }

    runLog.info("Workflow started", { attempt: run.attempts, payload: run.payload });
    try {
      await reg.handler(run.payload, context);
      await this.client.complete(run.id, this.workerId);
      runLog.info("Workflow completed");
    } catch (err) {
      if (err instanceof BailSignal) {
        await this.client.cancel(run.id, this.workerId, err.reason ?? null);
        runLog.info("Workflow cancelled", { reason: err.reason ?? null });
        return;
      }
      if (err instanceof FailSignal) {
        await this.client.fail(run.id, this.workerId, err.error.message, null, true);
        runLog.error("Workflow failed", { error: err.error });
        return;
      }
      if (err instanceof DeferSignal) {
        await this.client.defer(run.id, this.workerId, err.wakeAt);
        runLog.debug("Workflow deferred", { wakeAt: err.wakeAt });
        return;
      }
      if (err instanceof WaitSignal) {
        await this.client.waitForEvent(
          run.id,
          this.workerId,
          err.stepName,
          err.eventName,
          err.timeoutAt,
        );
        runLog.debug("Workflow waiting", { event: err.eventName, timeoutAt: err.timeoutAt });
        return;
      }

      // Regular error → run-level retry path.
      const message = err instanceof Error ? err.message : String(err);
      const maxAttempts = reg.retry?.maxAttempts ?? run.retries + 1;
      const finalize = run.attempts >= maxAttempts;
      const base = reg.retry?.backoff?.base ?? this.baseBackoffMs;
      const max = reg.retry?.backoff?.max ?? this.maxBackoffMs;
      const nextRunAt = finalize
        ? null
        : new Date(Date.now() + expBackoffMs(run.attempts, base, max));
      await this.client.fail(run.id, this.workerId, message, nextRunAt, finalize);
      const errField = err instanceof Error ? err : message;
      if (finalize) {
        runLog.error("Workflow failed", { attempt: run.attempts, error: errField });
      } else {
        runLog.warn("Workflow failed, retrying", {
          attempt: run.attempts,
          nextRunAt,
          error: errField,
        });
      }
    } finally {
      if (heartbeatTimer) clearInterval(heartbeatTimer);
    }
  }
}
